sklearn/datasets/openml.py

```diff
@@ -0,0 +1,415 @@
+import json
+import os
+from os.path import join, exists
+from warnings import warn
+
+try:
+    # Python 3+
+    from urllib.request import urlopen
+except ImportError:
+    # Python 2
+    from urllib2 import urlopen
+
+
+import numpy as np
+
+from sklearn.externals import arff
+from .base import get_data_home
+from ..externals.six import string_types, PY2
+from ..externals.six.moves.urllib.error import HTTPError
+from ..utils import Bunch, Memory
+
+__all__ = ['fetch_openml']
+
+
+_SEARCH_NAME = "https://openml.org/api/v1/json/data/list/data_name/{}/limit/1"
+_DATA_INFO = "https://openml.org/api/v1/json/data/{}"
+_DATA_FEATURES = "https://openml.org/api/v1/json/data/features/{}"
+_DATA_FILE = "https://openml.org/data/v1/download/{}"
+
+
+def _get_json_content_from_openml_api(url, error_message, raise_if_error):
+    """
+    Loads json data from the openml api
+
+    Parameters
+    ----------
+    url : str
+        The URL to load from. Should be an official OpenML endpoint
+
+    error_message : str or None
+        The error message to raise if an acceptable OpenML error is thrown
+        (acceptable error is, e.g., data id not found. Other errors, like 404's
+        will throw the native error message)
+
+    raise_if_error : bool
+        Whether to raise an error if OpenML returns an acceptable error (e.g.,
+        date not found). If this argument is set to False, a None is returned
+        in case of acceptable errors. Note that all other errors (e.g., 404)
+        will still be raised as normal.
+
+    Returns
+    -------
+    json_data : json or None
+        the json result from the OpenML server if the call was successful;
+        None otherwise iff raise_if_error was set to False and the error was
+        ``acceptable``
+    """
+    data_found = True
+    try:
+        response = urlopen(url)
+    except HTTPError as error:
+        # 412 is an OpenML specific error code, indicating a generic error
+        # (e.g., data not found)
+        if error.code == 412:
+            data_found = False
+        else:
+            raise error
+    if not data_found:
+        # not in except for nicer traceback
+        if raise_if_error:
+            raise ValueError(error_message)
```

Comment:
This function is the result of a refactoring of code by @amueller requested by @rth . The `error_message` that you point at is an expected error message, e.g., what comes up if the dataset does not exist. Specific http error messages are thrown a few lines above:
```
        if error.code == 412:
            data_found = False
        else:
            raise error
```
(which technically only can happen when the server is down or something similar)

Reply:
Okay. So there's no specific message in that HTTP response beyond the status code?

Reply:
Actually, yes there is. But realistically, it can only be one message (i.e., dataset by whichever id / name / version / activation_status combination not found)